mitmproxy/optmanager.py

```diff
@@ -1,37 +1,84 @@
 import contextlib
 import blinker
 import pprint
-import inspect
 import copy
 import functools
 import weakref
 import os
+import typing
+import textwrap
 
 import ruamel.yaml
 
 from mitmproxy import exceptions
 from mitmproxy.utils import typecheck
 
-
 """
     The base implementation for Options.
 """
 
+unset = object()
+
+
+class _Option:
+    __slots__ = ("name", "typespec", "value", "_default", "choices", "help")
+
+    def __init__(
+        self,
+        name: str,
+        default: typing.Any,
+        typespec: typing.Type,
+        help: str,
+        choices: typing.Optional[typing.Sequence[str]]
+    ) -> None:
+        typecheck.check_type(name, default, typespec)
+        self.name = name
+        self._default = default
+        self.typespec = typespec
+        self.value = unset
+        self.help = help
+        self.choices = choices
+
+    def __repr__(self):
+        return "{value} [{type}]".format(value=self.current(), type=self.typespec)
+
+    @property
+    def default(self):
+        return copy.deepcopy(self._default)
```

Comment:
Let's think about this. Immutable datastructures are great, but are often not nice to use in Python specifically.